## Práctica

Puedes utilizar este notebook para analisar el dataset propuesto

In [170]:

# Tu código a partir de aquí
import pandas as pd

df = pd.read_csv('../data/raw/p4ds-dataset-bancario.csv', sep=None, engine='python')
df.head()

,Fecha,Número ID,Sucursal,Producto,Monto USD,Vendedor,Estado App,Fecha Registro en la App,estado_mora
0,23/05/2025,1,Durazno,Premier,NaN,NaN,Registrado,24/05/2025,Cobrado
1,14/05/2025,2,Canelones,Préstamo Consumo,"672,0",C2,No Registrado,NaN,Cobrado
2,10/10/2025,3,Montevideo,Tarjeta de Crédito,NaN,NaN,No Registrado,NaN,En Mora
3,15/12/2025,4,Montevideo,Premier,NaN,NaN,Registrado,21/12/2025,Cobrado
4,30/12/2025,5,Montevideo,Préstamo Automotor,"20053,0",M1,Registrado,31/12/2025,Incobrable


In [171]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Fecha                     550 non-null    str  
 1   Número ID                 550 non-null    int64
 2   Sucursal                  550 non-null    str  
 3   Producto                  550 non-null    str  
 4   Monto USD                 225 non-null    str  
 5   Vendedor                  225 non-null    str  
 6   Estado App                550 non-null    str  
 7   Fecha Registro en la App  301 non-null    str  
 8   estado_mora               550 non-null    str  
dtypes: int64(1), str(8)
memory usage: 72.4 KB


In [172]:
df.describe()

,Número ID
count,550.000000
mean,275.500000
std,158.915596
min,1.000000
25%,138.250000
50%,275.500000
75%,412.750000
max,550.000000


In [173]:
df['Producto'].value_counts()

Producto
Cuenta Sueldo           157
Préstamo Consumo        145
Tarjeta de Crédito      122
Préstamo Automotor       49
Premier                  46
Préstamo Hipotecario     31
Name: count, dtype: int64

In [174]:
df['estado_mora'].value_counts()

estado_mora
Cobrado       396
En Mora        99
Incobrable     55
Name: count, dtype: int64

In [175]:
df['Sucursal'].value_counts()

Sucursal
Montevideo        127
Canelones         101
Maldonado          38
Colonia            36
Paysandú           30
Rivera             24
Durazno            20
Salto              19
Rocha              19
Tacuarembó         19
Río Negro          18
Cerro Largo        18
Florida            16
Flores             13
San José           13
Treinta y Tres     13
Lavalleja          10
Artigas             9
Soriano             7
Name: count, dtype: int64

In [176]:
df['Monto USD'].value_counts()

Monto USD
586,0      2
3826,0     2
3014,0     2
2329,0     2
672,0      1
          ..
97393,0    1
1966,0     1
2759,0     1
1538,0     1
2440,0     1
Name: count, Length: 221, dtype: int64

In [177]:
df['Monto USD'].value_counts().head(20)

Monto USD
586,0       2
3826,0      2
3014,0      2
2329,0      2
672,0       1
20053,0     1
2957,0      1
244058,0    1
3996,0      1
215753,0    1
9142,0      1
15672,0     1
885,0       1
117483,0    1
1069,0      1
4430,0      1
56581,0     1
3761,0      1
158056,0    1
2346,0      1
Name: count, dtype: int64

In [178]:
df['Fecha'].value_counts().head(20)

Fecha
14/02/2025    8
24/10/2025    7
27/01/2025    6
09/12/2025    6
09/07/2025    6
26/11/2025    6
08/01/2025    6
24/01/2025    5
15/10/2025    5
03/12/2025    5
22/05/2025    5
15/09/2025    5
04/11/2025    5
23/04/2025    5
11/08/2025    5
10/07/2025    5
10/10/2025    4
04/06/2025    4
02/06/2025    4
24/11/2025    4
Name: count, dtype: int64

In [179]:
df['Fecha Registro en la App'].value_counts()

Fecha Registro en la App
31/01/2025    4
31/10/2025    4
23/12/2025    3
22/07/2025    3
20/10/2025    3
             ..
06/04/2025    1
01/04/2025    1
17/08/2025    1
18/05/2025    1
09/09/2025    1
Name: count, Length: 213, dtype: int64

In [180]:
df['Fecha Registro en la App'].value_counts().head(20)

Fecha Registro en la App
31/01/2025    4
31/10/2025    4
23/12/2025    3
22/07/2025    3
20/10/2025    3
21/02/2025    3
23/05/2025    3
10/12/2025    3
21/07/2025    3
02/06/2025    3
01/12/2025    3
28/03/2025    3
24/08/2025    3
13/01/2025    3
09/11/2025    3
25/03/2025    3
24/05/2025    2
27/08/2025    2
09/06/2025    2
29/12/2025    2
Name: count, dtype: int64

## Carga y Exploración del Dataset

- Se carga el dataset bancario.
- El dataset contiene 550 filas y 9 columnas con información de productos bancarios de clientes.
- La columna `estado de mora` fue agregada por el profesor, no tiene sentido para todos los productos (ej: Cuenta Sueldo o Cuenta Premier), pero es interesante para los préstamos y las tarjetas de crédito.

## A corregir:
- `Fecha`: está como texto, hay que convertir a tipo fecha.
- `Fecha Registro en la App`: está como texto, hay que convertir a tipo fecha.
- `Monto USD`: esta como texto, hay que convertir a numérico.

## Nulos:
- `Monto USD` y `Vendedor` tienen nulos porque solo aplican a productos de préstamos hipotecarios, consumo y automotores.
- `Fecha Registro en la App` tiene nulos porque no todos los clientes están registrados en la App.



## Data Wrangling

In [181]:
df['Fecha'] = pd.to_datetime(df['Fecha'], format='%d/%m/%Y')


In [182]:
df['Fecha Registro en la App'] = pd.to_datetime(df['Fecha Registro en la App'], format='%d/%m/%Y')

In [183]:
df.dtypes

Fecha                       datetime64[us]
Número ID                            int64
Sucursal                               str
Producto                               str
Monto USD                              str
Vendedor                               str
Estado App                             str
Fecha Registro en la App    datetime64[us]
estado_mora                            str
dtype: object

In [184]:
df['Monto USD'].dtype

<StringDtype(na_value=nan)>

In [185]:
df.dtypes

Fecha                       datetime64[us]
Número ID                            int64
Sucursal                               str
Producto                               str
Monto USD                              str
Vendedor                               str
Estado App                             str
Fecha Registro en la App    datetime64[us]
estado_mora                            str
dtype: object

In [186]:
df['Monto USD'] = df['Monto USD'].astype(str).str.replace(',', '.').str.strip()
df['Monto USD'] = pd.to_numeric(df['Monto USD'], errors='coerce')
df['Monto USD'].dtype

dtype('float64')

In [187]:
df.dtypes

Fecha                       datetime64[us]
Número ID                            int64
Sucursal                               str
Producto                               str
Monto USD                          float64
Vendedor                               str
Estado App                             str
Fecha Registro en la App    datetime64[us]
estado_mora                            str
dtype: object

In [188]:
productos_prestamo = ['Préstamo Consumo', 'Préstamo Automotor', 'Préstamo Hipotecario']

df.loc[~df['Producto'].isin(productos_prestamo), 'estado_mora'] ='N/A'

df['estado_mora'].value_counts(dropna=False)

estado_mora
N/A           325
Cobrado       158
En Mora        43
Incobrable     24
Name: count, dtype: int64

La columna `estado_mora` fue relegada a productos de préstamos (Préstamo Consumo, Préstamo Automotor, Préstamo Hipotecario) ya que solo
tiene sentido de negocio para productos crediticios. Para el resto de los productos se asignó N/A.

In [189]:
df.loc[~df['Producto'].isin(productos_prestamo), 'Vendedor'] = 'N/A'

df['Vendedor'].value_counts(dropna=False).head(10)

Vendedor
N/A    325
M2      21
C2      19
M1      19
C1      14
M3      12
E2       7
C3       7
R2       6
S1       6
Name: count, dtype: int64

## Grabado del nuevo DataFrame en un archivo CSV

In [190]:
print(df['Monto USD'].dtype)
print(df['Monto USD'].head(10))
df.to_csv('../data/processed/p4ds-dataset-bancario-limpio.csv', index=False)
print("Dataset guardado correctamente!")

float64
0        NaN
1      672.0
2        NaN
3        NaN
4    20053.0
5     2957.0
6        NaN
7        NaN
8        NaN
9        NaN
Name: Monto USD, dtype: float64
Dataset guardado correctamente!
